examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,1162 @@
+embedded migration macro.
+
+To serve both purposes, deployment and CI testing, let's add a new function `run_db_migration` to the lib.rs file of the
+crate that takes a connection as parameter, checks if the DB connection is valid, checks if there are any pending
+migrations, and if so, runs all pending migrations. The implementation is straight forward, as you can see below:
+
+```rust 
+pub type Connection = PgConnection;
+
+pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!("migrations");
+
+pub fn run_db_migration(
+    conn: &mut Connection,
+) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+    // Check DB connection!
+    match conn.ping() {
+        Ok(_) => {}
+        Err(e) => {
+            eprint!("[run_db_migration]: Error connecting to database: {}", e);
+            return Err(Box::new(e));
+        }
+    }
+
+    // Check if DB has pending migrations
+    let has_pending = match conn.has_pending_migration(MIGRATIONS) {
+        Ok(has_pending) => has_pending,
+        Err(e) => {
+            eprint!(
+                "[run_db_migration]: Error checking for pending database migrations: {}",
+                e
+            );
+            return Err(e);
+        }
+    };
+
+    // If so, run all pending migrations.
+    if has_pending {
+        match conn.run_pending_migrations(MIGRATIONS) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                eprint!("[run_db_migration]: Error migrating database: {}", e);
+                Err(e)
+            }
+        }
+    } else {
+        // Nothing pending, just return
+        Ok(())
+    }
+}
+```
+
+The rational to check for all potential errors is twofold. For once, because the database connection is given as a
+parameter,
+you just don't know if the component that created the connection has checked the connection, therefore you better check
+it to catch a dead connection.
+Second, even if you have a correct database connection, this does not guarantee that the migration will succeed.
+There might be some types left from a previously aborted drop operations or
+a random error might happened at any stage of the migration, therefore you have to handle the error where it occurs.
+Also, because you run the db migration during application startup or before testing,
+ensure you have clear error messages to speed up diagnostic and debugging.
+
+### DB Integration Tests
+
+Database integration tests become flaky when executed in parallel usually because of conflicting read / write
+operations.
+While modern database systems can handle concurrent data access, test tools with assertions not so much.
+That means, test assertions start to fail seemingly randomly when executed concurrently.
+There are only very few viable options to deal with this reality:
+
+* Don't use parallel test execution
+* Only parallelize test per isolated access
+* Do synchronization and test the actual database state before each test and run tests as atomic transactions
```

Comment:
There is a simple trick here that enables you to write concurrent database tests: Use a dangling transaction. Diesel uses that extensively for it's own test-suite. What that means is that you perform the following steps:

* Open a new connection per test
* Directly start a new transaction (Diesel provides the `Connection::begin_test_transaction` function for this)
* Perform any test code (including schema setup for postgres)
* Rollback the transaction

With that you end up with mostly isolated views on the database, where each test does not see the changes from other tests. (Well you can observe different id's etc, but that's all)

Reply:
Okay, I actually added another testsuite to the example that implements the parallel testing via begin_test_transaction. That way, people have the freedom to adopt either the test group parallel test style or the per test transaction testing style.

Very good point, I've updated the Readme.